File: api/app/routers/health.py
"""
Health check endpoints.
- /health - Basic liveness check
- /ready - Readiness check with database connectivity
"""

from datetime import UTC, datetime

from fastapi import APIRouter, HTTPException

from ..config import get_settings
from ..db import get_pool
from ..models import HealthResponse, ReadyResponse

router = APIRouter(tags=["Health"])


@router.get("/health", response_model=HealthResponse)
async def health_check() -> HealthResponse:
    """
    Basic health check endpoint.
    Returns 200 if the service is running.
    """
    settings = get_settings()
    return HealthResponse(
        status="healthy",
        version=settings.app_version,
        timestamp=datetime.now(UTC),
    )


@router.get("/ready", response_model=ReadyResponse)
async def readiness_check() -> ReadyResponse:
    """
    Readiness check endpoint.
    Verifies database connectivity before returning healthy.
    """
    db_status = "unknown"

    try:
        pool = await get_pool()
        async with pool.acquire() as conn:
            result = await conn.fetchval("SELECT 1")
            if result == 1:
                db_status = "connected"
            else:
                db_status = "error"
    except Exception as e:
        db_status = f"error: {str(e)}"

    overall_status = "ready" if db_status == "connected" else "not_ready"

    if overall_status == "not_ready":
        raise HTTPException(
            status_code=503,
            detail=ReadyResponse(
                status=overall_status,
                database=db_status,
                timestamp=datetime.now(UTC),
            ).model_dump(),
        )

    return ReadyResponse(
        status=overall_status,
        database=db_status,
        timestamp=datetime.now(UTC),
    )

File: api-go/internal/handlers/health.go
// Package handlers provides HTTP request handlers.
package handlers

import (
	"net/http"
	"time"

	"github.com/gin-gonic/gin"
	"github.com/postgresql-ha-dr/api-go/internal/config"
	"github.com/postgresql-ha-dr/api-go/internal/db"
	"github.com/postgresql-ha-dr/api-go/internal/models"
)

// HealthHandler handles health check endpoints.
type HealthHandler struct {
	cfg  *config.Config
	pool *db.Pool
}

// NewHealthHandler creates a new health handler.
func NewHealthHandler(cfg *config.Config, pool *db.Pool) *HealthHandler {
	return &HealthHandler{
		cfg:  cfg,
		pool: pool,
	}
}

// Health handles GET /health - basic liveness check.
func (h *HealthHandler) Health(c *gin.Context) {
	c.JSON(http.StatusOK, models.HealthResponse{
		Status:    "healthy",
		Version:   h.cfg.App.Version,
		Timestamp: time.Now().UTC(),
	})
}

// Ready handles GET /ready - readiness check with database connectivity.
func (h *HealthHandler) Ready(c *gin.Context) {
	dbStatus := "unknown"

	if h.pool != nil {
		if err := h.pool.HealthCheck(c.Request.Context()); err != nil {
			dbStatus = "error: " + err.Error()
		} else {
			dbStatus = "connected"
		}
	} else {
		dbStatus = "not_initialized"
	}

	status := "ready"
	if dbStatus != "connected" {
		status = "not_ready"
	}

	response := models.ReadyResponse{
		Status:    status,
		Database:  dbStatus,
		Timestamp: time.Now().UTC(),
	}

	if status == "not_ready" {
		c.JSON(http.StatusServiceUnavailable, response)
		return
	}

	c.JSON(http.StatusOK, response)
}

// Root handles GET / - API info.
func (h *HealthHandler) Root(c *gin.Context) {
	c.JSON(http.StatusOK, gin.H{
		"message": "PostgreSQL HA/DR Demo API (Go)",
		"docs":    "/docs",
		"health":  "/health",
		"ready":   "/ready",
	})
}
